File: app/src/main/java/pl/srw/billcalculator/bill/calculation/PgeG12CalculatedBill.java
package pl.srw.billcalculator.bill.calculation;

import java.math.BigDecimal;

import lombok.Getter;
import pl.srw.billcalculator.pojo.IPgePrices;

/**
 * Created by Kamil Seweryn.
 */
@SuppressWarnings("FieldCanBeLocal")
@Getter
public class PgeG12CalculatedBill extends CalculatedEnergyBill {

    private final int dayConsumption;
    private final int nightConsumption;
    private final int dayConsumptionFromJuly16;
    private final int nightConsumptionFromJuly16;

    private final BigDecimal zaEnergieCzynnaDayNetCharge;
    private final BigDecimal skladnikJakosciowyDayNetCharge;
    private final BigDecimal oplataSieciowaDayNetCharge;
    private final BigDecimal zaEnergieCzynnaNightNetCharge;
    private final BigDecimal skladnikJakosciowyNightNetCharge;
    private final BigDecimal oplataSieciowaNightNetCharge;
    private final BigDecimal oplataOzeDayNetCharge;
    private final BigDecimal oplataOzeNightNetCharge;

    private final BigDecimal zaEnergieCzynnaDayVatCharge;
    private final BigDecimal skladnikJakosciowyDayVatCharge;
    private final BigDecimal oplataSieciowaDayVatCharge;
    private final BigDecimal zaEnergieCzynnaNightVatCharge;
    private final BigDecimal skladnikJakosciowyNightVatCharge;
    private final BigDecimal oplataSieciowaNightVatCharge;
    private final BigDecimal oplataOzeDayVatCharge;
    private final BigDecimal oplataOzeNightVatCharge;

    public PgeG12CalculatedBill(final int readingDayFrom, final int readingDayTo,
                                final int readingNightFrom, final int readingNightTo,
                                final String dateFrom, final String dateTo, final IPgePrices prices) {
        super(dateFrom, dateTo, prices.getOplataAbonamentowa(), prices.getOplataPrzejsciowa(), prices.getOplataStalaZaPrzesyl());
        dayConsumption = readingDayTo - readingDayFrom;
        nightConsumption = readingNightTo - readingNightFrom;
        dayConsumptionFromJuly16 = countConsumptionPartFromJuly16(dateFrom, dateTo, dayConsumption);
        nightConsumptionFromJuly16 = countConsumptionPartFromJuly16(dateFrom, dateTo, nightConsumption);

        zaEnergieCzynnaDayNetCharge = countNetAndAddToSum(prices.getZaEnergieCzynnaDzien(), dayConsumption);
        skladnikJakosciowyDayNetCharge = countNetAndAddToSum(prices.getSkladnikJakosciowy(), dayConsumption);
        oplataSieciowaDayNetCharge = countNetAndAddToSum(prices.getOplataSieciowaDzien(), dayConsumption);
        zaEnergieCzynnaNightNetCharge = countNetAndAddToSum(prices.getZaEnergieCzynnaNoc(), nightConsumption);
        skladnikJakosciowyNightNetCharge = countNetAndAddToSum(prices.getSkladnikJakosciowy(), nightConsumption);
        oplataSieciowaNightNetCharge = countNetAndAddToSum(prices.getOplataSieciowaNoc(), nightConsumption);
        oplataOzeDayNetCharge = countNetAndAddToSum(prices.getOplataOze(), dayConsumptionFromJuly16);
        oplataOzeNightNetCharge = countNetAndAddToSum(prices.getOplataOze(), nightConsumptionFromJuly16);

        zaEnergieCzynnaDayVatCharge = countVatAndAddToSum(zaEnergieCzynnaDayNetCharge);
        skladnikJakosciowyDayVatCharge = countVatAndAddToSum(skladnikJakosciowyDayNetCharge);
        oplataSieciowaDayVatCharge = countVatAndAddToSum(oplataSieciowaDayNetCharge);
        zaEnergieCzynnaNightVatCharge = countVatAndAddToSum(zaEnergieCzynnaNightNetCharge);
        skladnikJakosciowyNightVatCharge = countVatAndAddToSum(skladnikJakosciowyNightNetCharge);
        oplataSieciowaNightVatCharge = countVatAndAddToSum(oplataSieciowaNightNetCharge);
        oplataOzeDayVatCharge = countVatAndAddToSum(oplataOzeDayNetCharge);
        oplataOzeNightVatCharge = countVatAndAddToSum(oplataOzeNightNetCharge);
    }

    @Override
    public int getTotalConsumption() {
        return dayConsumption + nightConsumption;
    }
}

File: app/src/main/java/pl/srw/billcalculator/bill/calculation/PgeG12CalculatedBill.kt
package pl.srw.billcalculator.bill.calculation

import org.threeten.bp.LocalDate
import pl.srw.billcalculator.pojo.IPgePrices

class PgeG12CalculatedBill(
    readingDayFrom: Int,
    readingDayTo: Int,
    readingNightFrom: Int,
    readingNightTo: Int,
    dateFrom: LocalDate,
    dateTo: LocalDate,
    prices: IPgePrices
) : CalculatedEnergyBill(
    dateFrom,
    dateTo,
    prices.oplataAbonamentowa,
    prices.oplataPrzejsciowa,
    prices.oplataStalaZaPrzesyl,
    prices
) {

    val dayConsumption = readingDayTo - readingDayFrom
    val nightConsumption = readingNightTo - readingNightFrom
    val dayConsumptionFromJuly16 = countConsumptionPartFromJuly16(dateFrom, dateTo, dayConsumption)
    val nightConsumptionFromJuly16 = countConsumptionPartFromJuly16(dateFrom, dateTo, nightConsumption)

    val zaEnergieCzynnaDayNetCharge = countNetAndAddToSum(prices.zaEnergieCzynnaDzien, dayConsumption)
    val skladnikJakosciowyDayNetCharge = countNetAndAddToSum(prices.skladnikJakosciowy, dayConsumption)
    val oplataSieciowaDayNetCharge = countNetAndAddToSum(prices.oplataSieciowaDzien, dayConsumption)
    val zaEnergieCzynnaNightNetCharge = countNetAndAddToSum(prices.zaEnergieCzynnaNoc, nightConsumption)
    val skladnikJakosciowyNightNetCharge = countNetAndAddToSum(prices.skladnikJakosciowy, nightConsumption)
    val oplataSieciowaNightNetCharge = countNetAndAddToSum(prices.oplataSieciowaNoc, nightConsumption)
    val oplataOzeDayNetCharge = countNetAndAddToSum(prices.oplataOze, (dayConsumptionFromJuly16 * 0.001).toString())
    val oplataOzeNightNetCharge = countNetAndAddToSum(prices.oplataOze, (nightConsumptionFromJuly16 * 0.001).toString())

    val zaEnergieCzynnaDayVatCharge = countVatAndAddToSum(zaEnergieCzynnaDayNetCharge)
    val skladnikJakosciowyDayVatCharge = countVatAndAddToSum(skladnikJakosciowyDayNetCharge)
    val oplataSieciowaDayVatCharge = countVatAndAddToSum(oplataSieciowaDayNetCharge)
    val zaEnergieCzynnaNightVatCharge = countVatAndAddToSum(zaEnergieCzynnaNightNetCharge)
    val skladnikJakosciowyNightVatCharge = countVatAndAddToSum(skladnikJakosciowyNightNetCharge)
    val oplataSieciowaNightVatCharge = countVatAndAddToSum(oplataSieciowaNightNetCharge)
    val oplataOzeDayVatCharge = countVatAndAddToSum(oplataOzeDayNetCharge)
    val oplataOzeNightVatCharge = countVatAndAddToSum(oplataOzeNightNetCharge)

    override val totalConsumption = dayConsumption + nightConsumption
}
